File: src/Scoring.cpp
/*
 * Copyright 2020 Maximilian Azendorf
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "Scoring.h"

#include <cmath>

#include "Util.h"

bool Scoring::satisfies_constraints_scheduling(Scheduling const& scheduling)
{
    for(Constraint constraint : scheduling.input_data().scheduling_constraints())
    {
        int l = constraint.left();
        int r = constraint.right();
        int e = constraint.extra();

        switch(constraint.type())
        {
            case ChoiceIsInSlot:
                if(scheduling.slot_of(l) != r) return false;
                break;

            case ChoiceIsNotInSlot:
                if(scheduling.slot_of(l) == r) return false;
                break;

            case ChoicesAreInSameSlot:
                if(scheduling.slot_of(l) != scheduling.slot_of(r)) return false;
                break;

            case ChoicesAreNotInSameSlot:
                if(scheduling.slot_of(l) == Scheduling::NOT_SCHEDULED
                    || scheduling.slot_of(r) == Scheduling::NOT_SCHEDULED)
                {
                    break;
                }
                if(scheduling.slot_of(l) == scheduling.slot_of(r)) return false;
                break;

            case ChoicesHaveOffset:
                if(scheduling.slot_of(l) == Scheduling::NOT_SCHEDULED
                   || scheduling.slot_of(r) == Scheduling::NOT_SCHEDULED)
                {
                    break;
                }
                if(scheduling.slot_of(r) - scheduling.slot_of(l) != e) return false;
                break;

            case SlotHasLimitedSize:
            {
                int count = 0;
                for(int w = 0; w < scheduling.input_data().choice_count(); w++)
                {
                    if(scheduling.slot_of(w) == constraint.left())
                    {
                        count++;
                    }
                }

                switch((SlotSizeLimitOp)e)
                {
                    case Eq: if(count != r) return false; break;
                    case Neq: if(count == r) return false; break;
                    case Gt: if(count <= r) return false; break;
                    case Lt: if(count >= r) return false; break;
                    case Geq: if(count < r) return false; break;
                    case Leq: if(count > r) return false; break;
                    default: throw std::logic_error("Unknown slot size limit operator " + str(e) + ".");
                }
                break;
            }

            default: throw std::logic_error("Unknown scheduling constraint type " + str(constraint.type()) + ".");
        }
    }

    return true;
}

bool Scoring::satisfies_constraints_assignment(Scheduling const& scheduling, Assignment const& assignment)
{
    for(Constraint const& constraint : assignment.input_data().assignment_constraints())
    {
        int l = constraint.left();
        int r = constraint.right();
        //int e = constraint.extra();

        switch(constraint.type())
        {
            case ChoicesHaveSameChoosers:
                if(scheduling.slot_of(r) == Scheduling::NOT_SCHEDULED || scheduling.slot_of(l) == Scheduling::NOT_SCHEDULED) continue;
                if(assignment.choosers_ordered(l) != assignment.choosers_ordered(r)) return false;
                break;

            case ChooserIsInChoice:
                if(scheduling.slot_of(r) == Scheduling::NOT_SCHEDULED) continue;
                if(!assignment.is_in_choice(l, r)) return false;
                break;

            case ChooserIsNotInChoice:
                if(assignment.is_in_choice(l, r)) return false;
                break;

            case ChoosersHaveSameChoices:
                if(assignment.choices_ordered(l) != assignment.choices_ordered(r)) return false;
                break;

            default: throw std::logic_error("Unknown assignment constraint type " + str(constraint.type()) + ".");
        }
    }

    return true;
}

bool Scoring::satisfies_constraints(Solution const& solution) const
{
    return satisfies_constraints_scheduling(*solution.scheduling())
        && satisfies_constraints_assignment(*solution.scheduling(), *solution.assignment());
}

int Scoring::evaluate_major(Solution const& solution) const
{
    int m = 0;
    for(int i = 0; i < _inputData->chooser_count() * _inputData->slot_count(); i++)
    {
        int p = i / _inputData->slot_count();
        int ws = solution.assignment()->choice_of(p, i % _inputData->slot_count());
        m = std::max(m, _inputData->chooser(p).preferences[ws]);
    }

    return m;
}

float Scoring::evaluate_minor(Solution const& solution) const
{
    if(!is_feasible(solution))
    {
        return INFINITY;
    }

    vector<int> prefCount(_inputData->max_preference() + 1);

    for(int i = 0; i < _inputData->chooser_count() * _inputData->slot_count(); i++)
    {
        int p = i / _inputData->slot_count();
        int ws = solution.assignment()->choice_of(p, i % _inputData->slot_count());
        prefCount[_inputData->chooser(p).preferences[ws]]++;
    }

    float sum = 0;
    for(int pref = 0; pref <= _inputData->max_preference(); pref++)
    {
        sum += prefCount[pref] * std::pow((float)pref, (float)_options->preference_exponent()) / _scaling;
    }

    return sum;
}

Scoring::Scoring(const_ptr<InputData> inputData, const_ptr<Options> options)
        : _inputData(std::move(inputData)),
        _options(std::move(options))
{
    _scaling = std::pow((float)_inputData->max_preference(), (float)_options->preference_exponent());
}

bool Scoring::is_feasible(Solution const& solution) const
{
    vector<int> partCounts(_inputData->choice_count());
    vector<vector<bool>> isInSlot(_inputData->chooser_count(), vector<bool>(_inputData->slot_count()));
    vector<int> slots(_inputData->choice_count());

    if(!satisfies_constraints(solution))
    {
        return false;
    }

    for(int i = 0; i < _inputData->choice_count(); i++)
    {
        slots[i] = solution.scheduling()->slot_of(i);
    }

    for(int i = 0; i < _inputData->chooser_count() * _inputData->slot_count(); i++)
    {
        int p = i / _inputData->slot_count();
        int ws = solution.assignment()->choice_of(p, i % _inputData->slot_count());
        if(isInSlot[p][slots[ws]])
        {
            return false;
        }

        isInSlot[p][slots[ws]] = true;
        partCounts[ws]++;
    }

    for(int i = 0; i < _inputData->choice_count(); i++)
    {
        if (solution.scheduling()->slot_of(i) == Scheduling::NOT_SCHEDULED)
        {
            continue;
        }

        if(partCounts[i] < _inputData->choice(i).min || partCounts[i] > _inputData->choice(i).max)
        {
            return false;
        }
    }

    return true;
}

Score Scoring::evaluate(Solution const& solution) const
{
    if(solution.is_invalid())
    {
        return {.major = INFINITY, .minor = INFINITY};
    }

    auto major = _options->greedy() ? NAN : (float)evaluate_major(solution);
    auto minor = (float)evaluate_minor(solution);

    if((std::isfinite(major) || std::isnan(major)) && std::isfinite(minor))
    {
        return {.major = major, .minor = minor};
    }
    else
    {
        return {.major = INFINITY, .minor = INFINITY};
    }
}

File: src/scoring.rs
use std::sync::Arc;

use crate::{Assignment, InputData, Options, Scheduling, Score, Solution};

/// Evaluates complete solutions according to feasibility and preference quality.
#[derive(Debug, Clone)]
pub struct Scoring {
    input_data: Arc<InputData>,
    options: Arc<Options>,
    scaling: f32,
}

impl Scoring {
    /// Creates a new scorer for the given input and options.
    #[must_use]
    pub fn new(input_data: Arc<InputData>, options: Arc<Options>) -> Self {
        let scaling = (input_data.max_preference as f32).powf(options.preference_exponent as f32);
        Self {
            input_data,
            options,
            scaling,
        }
    }

    pub(crate) fn is_feasible(&self, solution: &Solution) -> bool {
        let scheduling = solution
            .scheduling()
            .expect("feasibility requires a scheduling");
        let assignment = solution
            .assignment()
            .expect("feasibility requires an assignment");

        let mut part_counts = vec![0_i32; self.input_data.choices.len()];
        let mut is_in_slot =
            vec![vec![false; self.input_data.slots.len()]; self.input_data.choosers.len()];
        let slots = (0..self.input_data.choices.len())
            .map(|choice| scheduling.slot_of(choice))
            .collect::<Vec<_>>();

        if !Self::satisfies_constraints_scheduling(scheduling)
            || !Self::satisfies_constraints_assignment(scheduling, assignment)
        {
            return false;
        }

        for (chooser, slots_in) in is_in_slot
            .iter_mut()
            .enumerate()
            .take(self.input_data.choosers.len())
        {
            for slot in 0..self.input_data.slots.len() {
                let choice = assignment.choice_of(chooser, slot);
                let scheduled_slot =
                    usize::try_from(slots[choice]).expect("choice must be scheduled");
                if slots_in[scheduled_slot] {
                    return false;
                }
                slots_in[scheduled_slot] = true;
                part_counts[choice] += 1;
            }
        }

        for (choice_index, &count) in part_counts.iter().enumerate() {
            if scheduling.slot_of(choice_index) == Scheduling::NOT_SCHEDULED {
                continue;
            }
            let choice = &self.input_data.choices[choice_index];
            if count < choice.min || count > choice.max {
                return false;
            }
        }

        true
    }

    /// Evaluates a solution and returns its score.
    #[must_use]
    pub fn evaluate(&self, solution: &Solution) -> Score {
        if solution.is_invalid() {
            return Score {
                major: f32::INFINITY,
                minor: f32::INFINITY,
            };
        }

        let major = if self.options.greedy {
            f32::NAN
        } else {
            self.evaluate_major(solution) as f32
        };
        let minor = self.evaluate_minor(solution);

        if (major.is_finite() || major.is_nan()) && minor.is_finite() {
            Score { major, minor }
        } else {
            Score {
                major: f32::INFINITY,
                minor: f32::INFINITY,
            }
        }
    }

    pub(crate) fn satisfies_constraints_scheduling(scheduling: &Scheduling) -> bool {
        for constraint in &scheduling.input_data.scheduling_constraints {
            let left = constraint.left;
            let right = usize::try_from(constraint.right).unwrap_or(usize::MAX);
            let extra = constraint.extra;

            match constraint.kind {
                crate::ConstraintType::ChoiceIsInSlot => {
                    if scheduling.slot_of(left) != constraint.right {
                        return false;
                    }
                }
                crate::ConstraintType::ChoiceIsNotInSlot => {
                    if scheduling.slot_of(left) == constraint.right {
                        return false;
                    }
                }
                crate::ConstraintType::ChoicesAreInSameSlot => {
                    if scheduling.slot_of(left) != scheduling.slot_of(right) {
                        return false;
                    }
                }
                crate::ConstraintType::ChoicesAreNotInSameSlot => {
                    if scheduling.slot_of(left) != Scheduling::NOT_SCHEDULED
                        && scheduling.slot_of(right) != Scheduling::NOT_SCHEDULED
                        && scheduling.slot_of(left) == scheduling.slot_of(right)
                    {
                        return false;
                    }
                }
                crate::ConstraintType::ChoicesHaveOffset => {
                    let left_slot = scheduling.slot_of(left);
                    let right_slot = scheduling.slot_of(right);
                    if left_slot == Scheduling::NOT_SCHEDULED
                        || right_slot == Scheduling::NOT_SCHEDULED
                    {
                        if left_slot != right_slot {
                            return false;
                        }
                    } else if right_slot - left_slot != extra {
                        return false;
                    }
                }
                crate::ConstraintType::SlotHasLimitedSize => {
                    let slot = i32::try_from(constraint.left).unwrap_or(i32::MAX);
                    let count = i32::try_from(
                        (0..scheduling.input_data.choices.len())
                            .filter(|&choice| scheduling.slot_of(choice) == slot)
                            .count(),
                    )
                    .unwrap_or(i32::MAX);
                    let valid = match extra {
                        1 => count == constraint.right,
                        -1 => count != constraint.right,
                        2 => count > constraint.right,
                        -3 => count < constraint.right,
                        3 => count >= constraint.right,
                        -2 => count <= constraint.right,
                        _ => panic!("Unknown slot size limit operator {extra}."),
                    };
                    if !valid {
                        return false;
                    }
                }
                _ => panic!("Unknown scheduling constraint type {:?}.", constraint.kind),
            }
        }

        true
    }

    pub(crate) fn satisfies_constraints_assignment(
        scheduling: &Scheduling,
        assignment: &Assignment,
    ) -> bool {
        for constraint in &assignment.input_data.assignment_constraints {
            let left = constraint.left;
            let right = usize::try_from(constraint.right).unwrap_or(usize::MAX);

            match constraint.kind {
                crate::ConstraintType::ChoicesHaveSameChoosers => {
                    if scheduling.slot_of(right) != Scheduling::NOT_SCHEDULED
                        && scheduling.slot_of(left) != Scheduling::NOT_SCHEDULED
                        && assignment.choosers_ordered(left) != assignment.choosers_ordered(right)
                    {
                        return false;
                    }
                }
                crate::ConstraintType::ChooserIsInChoice => {
                    if scheduling.slot_of(right) != Scheduling::NOT_SCHEDULED
                        && !assignment.is_in_choice(left, right)
                    {
                        return false;
                    }
                }
                crate::ConstraintType::ChooserIsNotInChoice => {
                    if assignment.is_in_choice(left, right) {
                        return false;
                    }
                }
                crate::ConstraintType::ChoosersHaveSameChoices => {
                    if assignment.choices_ordered(left) != assignment.choices_ordered(right) {
                        return false;
                    }
                }
                _ => panic!("Unknown assignment constraint type {:?}.", constraint.kind),
            }
        }

        true
    }
}

impl Scoring {
    fn evaluate_major(&self, solution: &Solution) -> i32 {
        let assignment = solution.assignment().expect("solution requires assignment");
        let mut max_pref = 0_i32;
        for chooser in 0..self.input_data.choosers.len() {
            for slot in 0..self.input_data.slots.len() {
                let choice = assignment.choice_of(chooser, slot);
                max_pref = max_pref.max(self.input_data.choosers[chooser].preferences[choice]);
            }
        }
        max_pref
    }

    fn evaluate_minor(&self, solution: &Solution) -> f32 {
        if !self.is_feasible(solution) {
            return f32::INFINITY;
        }

        let assignment = solution.assignment().expect("solution requires assignment");
        let mut pref_count =
            vec![0_i32; usize::try_from(self.input_data.max_preference + 1).expect("non-negative")];

        for chooser in 0..self.input_data.choosers.len() {
            for slot in 0..self.input_data.slots.len() {
                let choice = assignment.choice_of(chooser, slot);
                let pref = usize::try_from(self.input_data.choosers[chooser].preferences[choice])
                    .expect("preference must be non-negative");
                pref_count[pref] += 1;
            }
        }

        let mut sum = 0.0_f32;
        for (pref, &count) in pref_count.iter().enumerate() {
            sum += (count as f32) * (pref as f32).powf(self.options.preference_exponent as f32)
                / self.scaling;
        }
        sum
    }
}
